src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,90 +182,141 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
+                    bool targetNeedsRestore;
+                    ProjectLoadTelemetryReporter.TelemetryInfo targetTelemetryInfo;
+
+                    if (existingProjectTarget != null)
+                    {
+                        projectTargetsToUnload.Remove(existingProjectTarget);
+                        (targetTelemetryInfo, targetNeedsRestore) = await existingProjectTarget.UpdateWithNewProjectInfoAsync(loadedProjectInfo, hasAllInformation, _logger);
+                    }
+                    else
                     {
-                        AssemblyName = projectSystemName,
-                        FilePath = projectPath,
-                        CompilationOutputAssemblyFilePath = loadedProjectInfo.IntermediateOutputFilePath
-                    };
+                        // We haven't seen this target for this project before, so add it.
+                        var targetFramework = loadedProjectInfo.TargetFramework;
+                        var projectSystemName = targetFramework is null ? projectPath : $"{projectPath} (${targetFramework})";
 
-                    var projectSystemProject = await ProjectFactory.CreateAndAddToWorkspaceAsync(
-                        projectSystemName,
-                        loadedProjectInfo.Language,
-                        projectCreationInfo,
-                        _projectSystemHostInfo);
+                        var projectCreationInfo = new ProjectSystemProjectCreationInfo
+                        {
+                            AssemblyName = projectSystemName,
+                            FilePath = projectPath,
+                            CompilationOutputAssemblyFilePath = loadedProjectInfo.IntermediateOutputFilePath,
+                        };
 
-                    var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
-                    loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(projectToLoad with { ReportTelemetry = false });
-                    existingProjects.Add(loadedProject);
+                        var projectSystemProject = await ProjectFactory.CreateAndAddToWorkspaceAsync(
+                            projectSystemName,
+                            loadedProjectInfo.Language,
+                            projectCreationInfo,
+                            _projectSystemHostInfo);
 
-                    (targetTelemetryInfo, targetNeedsRestore) = await loadedProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                        var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
+                        existingProjectTargets.Add(loadedProject);
 
-                    needsRestore |= targetNeedsRestore;
```

Comment:
It looks like the original code was only propagating out "needs restore" in the case of a new LoadedProject being created. But it feels like we may need to restore a project that was already loaded (`if (existingProject != null)` above), if its PackageReferences change, for example. Was this code being only in this block intentional? @jasonmalinowski @dibarbet

Reply:
I adjusted the code to keep propagating out `targetNeedsRestore` under the same conditions, but make it more explicit that the value is dropped when the target already exists.

Reply:
Opened #78561. I feel the bug should be kept until the associated bug is fixed, in order to avoid nuisance toasts during ordinary source code editing.

Reply:
do file based programs need an actual `dotnet restore`?  Or does that happen in some other way?

Reply:
Yes. Our plan was to ensure SDK supports `dotnet restore app.cs`, and for VS Code to kick off a restore in a similar way as for ordinary projects.